### PR TITLE
feat(stella-model): streaming→non-streaming fallback on hung or empty streams, with a first-byte deadline (#2686)

### LINKED ISSUE #2686: IMPROVEMENT: streaming→non-streaming fallback on hung or empty streams, with a separate first-byte deadline
## The gap

Stella's streaming seam is `complete_observed_ref` with graceful *degradation* — an adapter without streaming just yields no speculation (`crates/stella-protocol/src/provider.rs:121-128`). But there's no *fallback*: when streaming itself is broken (proxy buffers the SSE body, gateway returns 200 with an empty stream, connection wedges mid-body), the idle deadline (816s of silence, `driver.rs:210-253`) eventually trips and is classified Terminal — never retried, turn aborted. A provider that would answer fine over a non-streaming request is treated as down. 816s is also a very long time to discover a proxy that buffers.

## What the comparator does

Zappy code detects two distinct broken-stream shapes early — a hung stream via an idle watchdog (90s default), and an **empty stream** (response completed with no content and no stop reason, a real proxy failure mode) — and falls back to a non-streaming request for the same payload. The hard part it solves is transcript hygiene: partially-streamed assistant messages (especially partial reasoning blocks, whose signatures are invalid) must be tombstoned/discarded before the retry, or the next request 400s; accumulated partial state (speculation pool, in-flight tool executor) is rebuilt.

## Proposed fix

On an idle-deadline trip or an empty-stream completion, retry the same attempt once via `complete_ref` (the non-streaming path already exists as the trait default). Discard partial attempt state — Stella's architecture makes this easier than zappy's: speculation discards are already fully accounted (`SpeculationDiscarded`), and the attempt hasn't been committed, so no transcript repair is needed beyond dropping the partial. Classify the streaming failure as retryable-once-via-fallback rather than Terminal, and consider a shorter first-byte deadline (distinct from the 816s inter-fragment deadline) to catch buffering proxies fast.

### PR DESCRIPTION
**Pull request by macanderson**
## What & why

A provider stream that hangs before its first byte (a proxy buffering the SSE body) or comes back as a 200 with an **empty stream** used to burn the whole retry budget re-issuing the identical streaming request into the same broken pipe: the per-chunk idle bound only fired after 120s per attempt, classified as retryable Transport, and every retry streamed again — a provider that would answer fine over a non-streaming request was treated as down. This is #2686's zappy-style "reactive resilience" fallback, adapted to Stella's accounting discipline.

Three pieces, all in `stella-model` (the engine is untouched):

- **A first-byte deadline, distinct from the idle bound.** `http::FIRST_BYTE_TIMEOUT` (90s, matching the comparator's stream watchdog) bounds only the FIRST body read; any chunk at all moves the stream onto the ordinary 120s `STREAM_IDLE_TIMEOUT`. It composes with — rather than duplicates — the engine's 816s idle-based model deadline: a hang now surfaces in 90s instead of 120s/816s. `http::next_stream_read` (the new `StreamRead` enum) keeps Idle distinct from Failed; `next_with_timeout` is reimplemented on top of it so there is one truth.
- **A latched, bounded recovery** (`src/stream_recovery.rs`): Streaming → Probing → Confirmed. A fallback-eligible fault (hung before the first byte, or the stream died having delivered *no completion signal whatsoever*) arms the probe; the probe's unary success confirms the latch for the session (a buffering proxy buffers deterministically); a failed probe **reverts** to streaming, so one unproven fault never condemns the streaming path. The retried attempt re-issues the byte-identical body with `stream: false` (`zai/unary.rs`) through `http::unary_client`'s 600s read bound (#547's lesson — a unary call has no first token to reset the clock).
- **Accounting rides the existing rails.** The recovery is deliberately split across two attempts rather than hidden inside one provider call: the faulted streaming attempt fails as retryable Transport, so the engine's retry ladder stays the single owner of retries and bills the discarded attempt through its existing per-attempt `UsageIncomplete` observer (`driver.rs::run_model_call`) — nothing arrived by construction of the eligibility rule, so `partial` is honestly `None`, exactly `http::partial_usage`'s "attach nothing rather than a zeroed envelope" discipline. A mid-stream death WITH salvage is deliberately ineligible: it keeps its retry-as-a-stream path and its `attach_partial` accounting, so there is no partial transcript to tombstone, ever.

**Invariant 8 (parity, declared not assumed):** the matrix gains its third axis, `StreamFallbackPosture`. The shared chat-completions adapter's identities (`zai`, `openrouter`, `xai`, `deepseek`, `local`, settings-defined gateways) declare `UnaryFallback` with named witnesses; `anthropic`/`openai`/`gemini`/`vertex` declare `StreamingOnly` with the gap tracked in #2746; `bedrock` is `AlwaysUnary` (it calls Converse — there is no stream to fall back from). Enforced from both sides like the other two axes: `stella-cli`'s config test fails a seeded provider with no row, and the parity tests fail a row whose witness rotted. AGENTS.md invariant 8 and the crate README are updated in the same PR (stale "two axes" prose is a bug).

**Exemplar:** the latch follows zappy's watchdog-and-fallback shape; the split-across-attempts structure follows this repo's own `retry.rs`/`UsageIncomplete` pattern rather than an internal hidden resend, so accounting and attempt-bounding stay owned by one place.

**God files:** `zai.rs` was at its 1565-line ceiling, so the SSE aggregation half moved to `zai/stream.rs` (the `driver/settlement.rs` split pattern) and the unary path landed in `zai/unary.rs`; `zai.rs` is now under 1500 and its baseline entry is retired (minimal baseline edit — deliberately NOT a full `file-size-update` regen, to avoid the parallel-merge retighten skew). `zai/tests.rs` was also at its ceiling, so the three `reasoning_aware_max_tokens` tests moved to `zai/tests/zai_effort.rs` (tests moved, not deleted — same names, new file). The streaming and unary paths share one set of assembly helpers (`tool_call_input`, `promote_reasoning_as_text`, `fold_usage`, `final_finish_reason`) so they cannot drift.

Closes #2686

## The witness

- [x] This PR includes a witness test (fails on `main`, passes here)

`crates/stella-model/src/zai/tests/stream_fallback.rs`:

- `a_stream_hung_before_its_first_byte_falls_back_to_a_non_streaming_request` — the headline #2686 witness. Wiremock cannot hold a response body open (its delays cover the whole response), so this one uses a minimal hand-rolled std-TCP server that answers a `"stream":true` request with headers and **not one body byte**, and a `"stream":false` request with a real completion. First call faults retryably at the first-byte deadline naming the switch; second call completes unary.
- `an_empty_stream_falls_back_to_a_non_streaming_request` — the other broken shape (200 + EOF before any data), wiremock-based; the unary response carries a tool call, proving the fallback parses the whole dialect.
- Controls: `a_healthy_stream_never_arms_the_fallback` (every request keeps `stream:true`), `a_stream_that_died_after_content_is_retried_as_a_stream_not_unary` (eligibility boundary), `a_failed_unary_probe_reverts_the_session_to_streaming` (wire sequence `stream → unary → stream` proves the reversion), plus latch state-machine units in `stream_recovery.rs`.

**Flip evidence, run the artisanal way** (`git checkout origin/main -- crates/stella-model`, inject the empty-stream witness, `cargo test -p stella-model an_empty_stream_falls_back`):

```
test zai::tests::an_empty_stream_falls_back_to_a_non_streaming_request ... FAILED
panicked at crates/stella-model/src/zai/tests.rs:1933:
provider transport error: Z.ai stream ended before [DONE] — connection closed mid-response
```

— on main the fault carries no fallback and the retry re-streams; with the change the same suite is green (382 passed). The hung-first-byte witness additionally cannot compile on main (`with_first_byte_deadline` and the latch don't exist).

## The gate

- [x] `cargo fmt --check`
- [x] `cargo clippy -p stella-model -p stella-cli --all-targets -- -D warnings` (exit 0)
- [x] `cargo test -p stella-model` (382 passed) and `stella-cli`'s seeded-provider parity tests (3 passed) after rebasing onto `origin/main` at #2720's ledger merge
- [x] `RUSTDOCFLAGS="-D warnings" cargo doc --no-deps -p stella-model -p stella-cli` — exit 0, checked unpiped
- [x] Docs updated: AGENTS.md invariant 8 (three axes), `crates/stella-model/README.md` (layout + god-file list), module docs
- [x] `Closes #2686` appears both here and as a commit trailer

## Nothing left behind

- #2746 — extend the fallback to the `anthropic`, `openai`, `gemini`/`vertex` dialects (written as a full handoff; the `StreamingOnly` parity rows cite it). Bedrock needs nothing.
- Not fixed here by design: a proxy that heartbeats keep-alive bytes forever while buffering the real body resets the idle clock and is not caught by either deadline — same as before this PR; noted in the aggregator's first-chunk comment and left out of #2746's scope deliberately (no such middlebox has been observed).

## Summary by Sourcery

Introduce a streaming→unary fallback in the shared OpenAI-compatible adapter when streams hang before the first byte or return empty, backed by a per-session latch and a first-byte timeout, while keeping retry and usage accounting unchanged.

New Features:
- Add a per-session streaming recovery latch that can reroute retried attempts from streaming to unary when a stream is hung or empty.
- Provide a non-streaming (unary) completion path for the shared chat-completions adapter that mirrors the streaming path’s assembly semantics.
- Define a first-byte timeout for streaming responses to detect hung streams separately from inter-chunk idle time.

Enhancements:
- Refactor the Zai adapter into separate streaming and unary modules sharing common assembly helpers to prevent divergence between delivery paths.
- Extend the provider parity matrix with a StreamFallbackPosture axis, declaring each provider’s streaming recovery behavior and enforcing completeness via tests.
- Add a generalized stream-read helper that distinguishes idle, end-of-stream, and transport failures for more precise handling of streaming faults.

Documentation:
- Update AGENTS.md and the stella-model README to describe the new StreamFallbackPosture axis and the split Zai streaming/unary modules, including stream-recovery and first-byte timeout plumbing.

Tests:
- Add a dedicated stream_fallback test suite for Zai covering hung-first-byte, empty-stream, healthy-stream control, mid-stream death, and failed unary probe behaviors.
- Update parity and CLI config tests to require StreamFallbackPosture rows and witness tests for providers that claim a unary fallback.
- Move reasoning-aware max-tokens tests into a separate zai_effort test module to stay within file-size limits.

### CODE REVIEW

File: crates/stella-model/src/zai/unary.rs [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+    /// probing confirms the fallback for the session, failure reverts to
+    /// streaming (see `crate::stream_recovery` for why neither is skipped).
+    pub(super) async fn complete_unary_attempt(
+        &self,
+        req: CompletionRequestRef<'_>,
+        force_default_reasoning: bool,
+    ) -> Result<CompletionResult, ProviderError> {
+        let body = self.build_body(req, force_default_reasoning, false);
+        let outcome = async {
+            let response = self.dispatch(&self.unary_client, &body).await?;
```
[the remark this answers is not in the document]
Comment (macanderson) [reply crates/stella-model/src/zai/unary.rs]: Verified against merged main (f83692c82) — the concern was real and both halves are now fixed:

- **`.send()` path**: fixed before merge. `crates/stella-model/src/zai.rs:1242` routes unary send errors through `http::classify_unary_dispatch_error` (read-timeout → non-retryable `Terminal`, connect → retryable; `http.rs:110-122`), witnessed by `a_unary_read_timeout_is_terminal_never_a_retry_storm` (`zai/tests/stream_fallback.rs`), which pins both directions — Terminal on the unary dispatch, still-retryable on the streaming dispatch, where the same expiry is only a header stall.
- **body read (`response.text()`)**: this comment's catch, applied at merge — `crates/stella-model/src/zai/unary.rs` (`complete_unary_attempt`) now maps `text()` errors through the same classifier, so a generation whose wait lands in the body (a gateway that sends early headers) cannot re-enter the #547 identical-request storm either.

The applied body-read hunk has no test pinning it (the existing witness only exercises `send()`, where wiremock's whole-response delay lands) — filed #2756 to witness it with a headers-then-stalled-body server.

Note for anyone landing follow-ups: main is currently red from an unrelated merge composition between this PR and #2752 (`provider_parity.rs::adapter_sources` — both entries survived, one array length). Repair is up as #2754.

**Comment by macanderson** [comment]
**Audit trail: tests renamed/moved by this PR** (the record `check-deleted-tests` asked for on the merge check — the PR merged before the description could carry it):

- **Renamed** (this is the one name the guard flagged): `both_axes_cover_the_same_provider_ids` → `all_axes_cover_the_same_provider_ids` in `crates/stella-model/src/provider_parity.rs`, widened to cover the new third axis (`StreamFallbackPosture`) alongside cache and reasoning.
- **Moved, names unchanged** (tree-wide the guard is quiet on these; listed for completeness): `an_uncapped_reasoning_turn_gets_thinking_headroom`, `a_pinned_cap_is_honored_whatever_the_reasoning_setting`, `a_non_reasoning_turn_still_sends_no_cap` — from `crates/stella-model/src/zai/tests.rs` to `crates/stella-model/src/zai/tests/zai_effort.rs` (the parent is a god file at its ceiling; the move made room for `mod stream_fallback;`).
- No test was deleted.

Also, for anyone reading that merge-check log: the `an_empty_stream_falls_back_to_a_non_streaming_request ... FAILED / panicked at zai/tests.rs:1933` lines inside the `check-deleted-tests` step are **not a test run** — they are this PR description's own flip-evidence block (the fail-on-main proof) echoed by the guard while scanning the PR body. On merged main the real witness passes: `cargo test -p stella-model` is green once #2754's `adapter_sources` repair lands (main is currently red from a #2748×#2752 merge composition unrelated to the fallback behavior).
